README.md
# i2c_adc_ads7828
[![GitHub release](https://img.shields.io/github/release/4-20ma/i2c_adc_ads7828.svg?maxAge=3600)][GitHub release]
[![Travis](https://img.shields.io/travis/4-20ma/i2c_adc_ads7828.svg?maxAge=3600)][Travis]
[![license](https://img.shields.io/github/license/4-20ma/i2c_adc_ads7828.svg?maxAge=3600)][license]
[![code of conduct](https://img.shields.io/badge/%E2%9D%A4-code%20of%20conduct-blue.svg?maxAge=3600)][code of conduct]

[GitHub release]:   https://github.com/4-20ma/i2c_adc_ads7828
[Travis]:           https://travis-ci.org/4-20ma/i2c_adc_ads7828
[license]:          LICENSE
[code of conduct]:  CODE_OF_CONDUCT.md


## Overview
This is an Arduino library for the Texas Instruments ADS7828 12-bit, 8-channel I<sup>2</sup>C A/D converter.


## Features
The following features are available:

  - Up to (4) A/D converters can be used on the same I<sup>2</sup>C bus (hardware-addressable via pins A0, A1 and software-addressable via ID 0..3; address 0x48..0x4C)
  - A/D conversions may be initiated on a bus-, device-, or channel-specific level
  - Retrieve values as 16-period moving average or last sample
  - Built-in scaling function to return values in user-defined engineering units


## Installation

#### Library Manager
Install the library into your Arduino IDE using the Library Manager (available from IDE version 1.6.2). Open the IDE and click Sketch > Include Library > Manage Libraries&hellip;

Scroll or search for `i2c_adc_ads7828`, then select the version of the library you want to install. Quit/re-launch the IDE to refresh the list; new versions are automatically added to the list, once released on GitHub.

Refer to Arduino Tutorials > Libraries [Using the Library Manager](https://www.arduino.cc/en/Guide/Libraries#toc3).

#### Zip Library
Refer to Arduino Tutorials > Libraries [Importing a .zip Library](https://www.arduino.cc/en/Guide/Libraries#toc4).

#### Manual
Refer to Arduino Tutorials > Libraries [Manual Installation](https://www.arduino.cc/en/Guide/Libraries#toc5).


## Schematic
This library has been tested with an Arduino [Duemilanove](http://www.arduino.cc/en/Main/ArduinoBoardDuemilanove) and a Texas Instruments [ADS7828](http://focus.ti.com/docs/prod/folders/print/ads7828.html) A/D converter.

Below is a simplified schematic diagram. Refer to the datasheet for specific requirements.

```
                               Arduino
                              .--------------------.
                              |Duemilanove         |
                              |                    |
   TI ADS7828        .-------o|5V                  |
  .--------------.   |        |                    |
-o|1 CH0   VDD 16|o--'   .---o|GND                 |
  |              |       |    |                    |
-o|2 CH1   SDA 15|o------)---o|A4 SDA              |
  |              |       |    |                    |
-o|3 CH2   SCL 14|o------)---o|A5 SCL              |
  |              |       |    '--------------------'
-o|4 CH3    A1 13|o------o
  |              |       |
-o|5 CH4    A0 12|o------o
  |              |       |
-o|6 CH5   COM 11|o-     |
  |              |       |
-o|7 CH6   REF 10|o-     |
  |              |       |
-o|8 CH7   GND  9|o------o
  '--------------'       |
                         |
                        ===
                        GND
```


## Example
The library contains sketches that demonstrates use of the `i2c_adc_ads7828` library. You can find these in the [examples](/examples/) folder.

``` cpp
#include <i2c_adc_ads7828.h>


// device 0
// Address: A1=0, A0=0
// Command: SD=1, PD1=1, PD0=1
ADS7828 device(0, SINGLE_ENDED | REFERENCE_ON | ADC_ON, 0x0F);
ADS7828* adc = &device;
ADS7828Channel* ambientTemp = adc->channel(0);
ADS7828Channel* waterTemp = adc->channel(1);
ADS7828Channel* filterPressure = adc->channel(2);
ADS7828Channel* waterLevel = adc->channel(3);


void setup()
{
  // enable serial monitor
  Serial.begin(19200);

  // enable I2C communication
  ADS7828::begin();

  // adjust scaling on an individual channel basis
  ambientTemp->minScale = 0;
  ambientTemp->maxScale = 150;

  waterTemp->minScale = 0;
  waterTemp->maxScale = 100;

  filterPressure->minScale = 0;
  filterPressure->maxScale = 30;

  waterLevel->minScale = 0;
  waterLevel->maxScale = 100;
}


void loop()
{
  // update all registered ADS7828 devices/unmasked channels
  ADS7828::updateAll();

  // output moving average values to console
  Serial.print("\n Ambient: ");
  Serial.print(ambientTemp->value(), DEC);
  Serial.print("\n Water temp: ");
  Serial.print(waterTemp->value(), DEC);
  Serial.print("\n Filter pressure: ");
  Serial.print(filterPressure->value(), DEC);
  Serial.print("\n Water level: ");
  Serial.print(waterLevel->value(), DEC);
  Serial.print("\n- - - - - - - - - - - - - - - - - - - - \n");

  // delay
  delay(1000);
}
```


## Caveats
Conforms to Arduino IDE 1.5 Library Specification v2.1 which requires Arduino IDE >= 1.5.


## Support
Please [submit an issue](https://github.com/4-20ma/i2c_adc_ads7828/issues) for all questions, bug reports, and feature requests. Email requests will be politely redirected to the issue tracker so others may contribute to the discussion and requestors get a more timely response.


## License & Authors

- Author:: Doc Walker ([4-20ma@wvfans.net](mailto:4-20ma@wvfans.net))

```
Copyright:: 2009-2016 Doc Walker

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# i2c_adc_ads7828 CHANGELOG

## [v2.0.1](https://github.com/4-20ma/i2c_adc_ads7828/tree/v2.0.1) (2016-09-27)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v2.0.0...v2.0.1)

BUG FIXES

- Fix Rakefile git remote/branch [\#22](https://github.com/4-20ma/i2c_adc_ads7828/pull/22) ([4-20ma](https://github.com/4-20ma))

## [v2.0.0](https://github.com/4-20ma/i2c_adc_ads7828/tree/v2.0.0) (2016-09-27)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.1.5...v2.0.0)

IMPROVEMENTS

- Add Code of Conduct [\#16](https://github.com/4-20ma/i2c_adc_ads7828/pull/16) ([4-20ma](https://github.com/4-20ma))
- \[BREAK\] Change project license to Apache 2.0 [\#8](https://github.com/4-20ma/i2c_adc_ads7828/pull/8) ([4-20ma](https://github.com/4-20ma))
- Automate CHANGELOG [\#6](https://github.com/4-20ma/i2c_adc_ads7828/pull/6) ([4-20ma](https://github.com/4-20ma))
- Add continuous integration testing via Travis CI [\#4](https://github.com/4-20ma/i2c_adc_ads7828/pull/4) ([4-20ma](https://github.com/4-20ma))
- \[BREAK\] Update to IDE 1.5 library format v2.1 [\#3](https://github.com/4-20ma/i2c_adc_ads7828/pull/3) ([4-20ma](https://github.com/4-20ma))

OTHER

- Fix README link to examples folder [\#20](https://github.com/4-20ma/i2c_adc_ads7828/pull/20) ([4-20ma](https://github.com/4-20ma))
- Remove extraneous \#includes/whitespace [\#18](https://github.com/4-20ma/i2c_adc_ads7828/pull/18) ([4-20ma](https://github.com/4-20ma))
- Update README content, layout, extension [\#14](https://github.com/4-20ma/i2c_adc_ads7828/pull/14) ([4-20ma](https://github.com/4-20ma))
- Add standard files to .gitignore [\#12](https://github.com/4-20ma/i2c_adc_ads7828/pull/12) ([4-20ma](https://github.com/4-20ma))
- Rename .pde files .ino [\#10](https://github.com/4-20ma/i2c_adc_ads7828/pull/10) ([4-20ma](https://github.com/4-20ma))

## [v1.1.5](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.1.5) (2012-12-29)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.1.4...v1.1.5)

## [v1.1.4](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.1.4) (2012-12-29)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.1.3...v1.1.4)

## [v1.1.3](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.1.3) (2012-12-29)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.1.2...v1.1.3)

## [v1.1.2](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.1.2) (2012-12-29)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.1.1...v1.1.2)

## [v1.1.1](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.1.1) (2012-12-29)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.1.0...v1.1.1)

## [v1.1.0](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.1.0) (2012-12-29)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.0.2...v1.1.0)

## [v1.0.2](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.0.2) (2012-01-08)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.0.1...v1.0.2)

## [v1.0.1](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.0.1) (2012-01-07)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v1.0...v1.0.1)

## [v1.0](https://github.com/4-20ma/i2c_adc_ads7828/tree/v1.0) (2012-01-07)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.7...v1.0)

## [v0.7](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.7) (2011-12-27)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.6...v0.7)

## [v0.6](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.6) (2011-12-27)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.5...v0.6)

## [v0.5](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.5) (2011-12-27)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.4...v0.5)

## [v0.4](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.4) (2010-02-10)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.3...v0.4)

## [v0.3](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.3) (2010-02-07)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.2...v0.3)

## [v0.2](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.2) (2010-02-07)
[Full Changelog](https://github.com/4-20ma/i2c_adc_ads7828/compare/v0.1...v0.2)

## [v0.1](https://github.com/4-20ma/i2c_adc_ads7828/tree/v0.1) (2010-01-31)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

i2c_adc_ads7828.h
/// \file
/// Arduino library for TI ADS7828 I2C A/D converter.
/*

  i2c_adc_ads7828.h - Arduino library for TI ADS7828 I2C A/D converter
  
  This file is part of i2c_adc_ads7828.
  
  i2c_adc_ads7828 is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  i2c_adc_ads7828 is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with i2c_adc_ads7828.  If not, see <http://www.gnu.org/licenses/>.
  
  Written by Doc Walker (Rx)
  Copyright © 2009-2012 Doc Walker <4-20ma at wvfans dot net>
  
*/


/// \mainpage Arduino library for TI ADS7828 I2C A/D converter.
/// 
/// The ADS7828 is a single-supply, low-power, 12-bit data acquisition 
/// device that features a serial I2C interface and an 8-channel 
/// multiplexer. The Analog-to-Digital (A/D) converter features a 
/// sample-and-hold amplifier and internal, asynchronous clock. The 
/// combination of an I2C serial, 2-wire interface and micropower 
/// consumption makes the ADS7828 ideal for applications requiring the A/D 
/// converter to be close to the input source in remote locations and for 
/// applications requiring isolation. The ADS7828 is available in a TSSOP-16 
/// package. 
/// \author Doc Walker
/// \version \verbinclude VERSION
/// \date 29 Dec 2012
/// \copyright GNU General Public License v3
/// \par Source Code Repository:
///   https://github.com/4-20ma/i2c_adc_ads7828
/// \par Programming Style Guidelines:
///   http://geosoft.no/development/cppstyle.html
/// \par Schematic:
///   \verbinclude SCHEMATIC


#ifndef i2c_adc_ads7828_h
#define i2c_adc_ads7828_h

// _________________________________________________________ STANDARD INCLUDES
// include types & constants of Wiring core API
#if defined(ARDUINO) && ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif


// __________________________________________________________ PROJECT INCLUDES
// include twi/i2c library
#include <Wire.h>


// ____________________________________________________________ UTILITY MACROS


// _________________________________________________________________ CONSTANTS
/// Configure channels to use differential inputs (Command byte SD=0).
/// Use either \ref DIFFERENTIAL or \ref SINGLE_ENDED in ADS7828
///   constructor; default is \ref DIFFERENTIAL.
/// \par Usage:
/// \code
/// ...
/// // address 0, differential inputs, ref/ADC OFF between conversions
/// ADS7828 adc0(0, DIFFERENTIAL | REFERENCE_OFF | ADC_OFF);
/// ...
/// \endcode
/// \relates ADS7828
static const uint8_t DIFFERENTIAL        = 0 << 7; // SD == 0


/// Configure channels to use single-ended inputs (Command byte SD=1).
/// Use either \ref DIFFERENTIAL or \ref SINGLE_ENDED in ADS7828
///   constructor; default is \ref DIFFERENTIAL.
/// \par Usage:
/// \code
/// ...
/// // address 1, single-ended inputs, ref/ADC OFF between conversions
/// ADS7828 adc1(1, SINGLE_ENDED | REFERENCE_OFF | ADC_OFF);
/// ...
/// \endcode
/// \relates ADS7828
static const uint8_t SINGLE_ENDED         = 1 << 7; // SD == 1


/// Configure channels to turn internal reference OFF between conversions (Command byte PD1=0).
/// Use either \ref REFERENCE_OFF or \ref REFERENCE_ON in ADS7828
///   constructor; default is \ref REFERENCE_OFF.
/// \par Usage:
/// \code
/// ...
/// // address 0, differential inputs, ref/ADC OFF between conversions
/// ADS7828 adc0(0, DIFFERENTIAL | REFERENCE_OFF | ADC_OFF);
/// ...
/// \endcode
/// \relates ADS7828
static const uint8_t REFERENCE_OFF        = 0 << 3; // PD1 == 0


/// Configure channels to turn internal reference ON between conversions (Command byte PD1=1).
/// Use either \ref REFERENCE_OFF or \ref REFERENCE_ON in ADS7828
///   constructor; default is \ref REFERENCE_OFF.
/// \par Usage:
/// \code
/// ...
/// // address 2, differential inputs, ref ON/ADC OFF between conversions
/// ADS7828 adc2(2, DIFFERENTIAL | REFERENCE_ON | ADC_OFF);
/// ...
/// \endcode
/// \relates ADS7828
static const uint8_t REFERENCE_ON         = 1 << 3; // PD1 == 1


/// Configure channels to turn A/D converter OFF between conversions (Command byte PD0=0).
/// Use either \ref ADC_OFF or \ref ADC_ON in ADS7828
///   constructor; default is \ref ADC_OFF.
/// \par Usage:
/// \code
/// ...
/// // address 0, differential inputs, ref/ADC OFF between conversions
/// ADS7828 adc0(0, DIFFERENTIAL | REFERENCE_OFF | ADC_OFF);
/// ...
/// \endcode
/// \relates ADS7828
static const uint8_t ADC_OFF              = 0 << 2; // PD0 == 0


/// Configure channels to turn A/D converter ON between conversions (Command byte PD0=1).
/// Use either \ref ADC_OFF or \ref ADC_ON in ADS7828
///   constructor; default is \ref ADC_OFF.
/// \par Usage:
/// \code
/// ...
/// // address 3 , differential inputs, ref OFF/ADC ON between conversions
/// ADS7828 adc3(3, DIFFERENTIAL | REFERENCE_OFF | ADC_ON);
/// ...
/// \endcode
/// \relates ADS7828
static const uint8_t ADC_ON               = 1 << 2; // PD0 == 1


/// Default channel mask used in ADS7828 constructor.
/// \relates ADS7828
static const uint8_t DEFAULT_CHANNEL_MASK  = 0xFF;


/// Default scaling minimum value used in ADS7828 constructor.
/// \relates ADS7828
static const uint16_t DEFAULT_MIN_SCALE    = 0;


/// Default scaling maximum value used in ADS7828 constructor.
/// \relates ADS7828
static const uint16_t DEFAULT_MAX_SCALE    = 0xFFF;


// _________________________________________________________ CLASS DEFINITIONS
class ADS7828;
class ADS7828Channel
{
  public:
    // ............................................... public member functions
    ADS7828Channel() {};
    ADS7828Channel(ADS7828* const, uint8_t, uint8_t, uint16_t, uint16_t);
    uint8_t commandByte();
    ADS7828* device();
    uint8_t id();
    uint8_t index();
    void newSample(uint16_t);
    void reset();
    uint16_t sample();
    uint8_t start();
    uint16_t total();
    uint8_t update();
    uint16_t value();
    
    // ........................................ static public member functions
    
    // ..................................................... public attributes
    /// Maximum value of moving average (defaults to 0x0FFF).
    /// \par Usage:
    /// \code
    /// ...
    /// ADS7828 device(0);
    /// ADS7828Channel* temperature = device.channel(0);
    /// uint16_t old = temperature->maxScale; // get current value and/or
    /// temperature->maxScale = 100;          // set new value
    /// ...
    /// \endcode
    uint16_t maxScale;
    
    /// Minimum value of moving average (defaults to 0x0000).
    /// \par Usage:
    /// \code
    /// ...
    /// ADS7828 device(0);
    /// ADS7828Channel* temperature = device.channel(0);
    /// uint16_t old = temperature->minScale; // get current value and/or
    /// temperature->minScale = 0;            // set new value
    /// ...
    /// \endcode
    uint16_t minScale;
    
    // .............................................. static public attributes
  
  private:
    // .............................................. private member functions
    
    // ....................................... static private member functions
    
    // .................................................... private attributes
    /// Command byte for channel object (SD C2 C1 C0 bits only).
    uint8_t commandByte_;
    
    /// Pointer to parent device object.
    ADS7828* device_;
    
    /// Index position within moving average array. 
    uint8_t index_;
    
    /// Array of (unscaled) sample values.
    /// \note Bit shift must match \ref MOVING_AVERAGE_BITS_.
    uint16_t samples_[1 << 4];
    
    /// (Unscaled) running total of moving average array elements.
    uint16_t total_;
    
    // ............................................. static private attributes
    /// Quantity of samples to be averaged =
    ///   2<sup>\ref MOVING_AVERAGE_BITS_</sup>.
    /// \note \ref MOVING_AVERAGE_BITS_ must match \ref samples_ bit shift.
    static const uint8_t MOVING_AVERAGE_BITS_ = 4;
};


class ADS7828
{
  public:
    // ............................................... public member functions
    ADS7828(uint8_t);
    ADS7828(uint8_t, uint8_t);
    ADS7828(uint8_t, uint8_t, uint8_t);
    ADS7828(uint8_t, uint8_t, uint8_t, uint16_t, uint16_t);
    uint8_t address();
    ADS7828Channel* channel(uint8_t);
    uint8_t commandByte();
    uint8_t start();
    uint8_t start(uint8_t);
    uint8_t update(); // single device, all unmasked channel
    uint8_t update(uint8_t); // single device, single channel
    
    // ........................................ static public member functions
    static void begin();
    static ADS7828* device(uint8_t);
    static uint8_t updateAll(); // all devices, all unmasked channels
    
    // ..................................................... public attributes
    /// Each bit position containing a 1 represents a channel that is to be
    /// read via update() / updateAll().
    uint8_t channelMask;                    // mask of active channels
    
    // .............................................. static public attributes
  
  private:
    // .............................................. private member functions
    void init(uint8_t, uint8_t, uint8_t, uint16_t, uint16_t);
    uint16_t read();
    
    // ....................................... static private member functions
    static uint16_t read(uint8_t);
    static uint8_t start(uint8_t, uint8_t);
    static uint8_t update(ADS7828*); // single device, all unmasked channels
    static uint8_t update(ADS7828*, uint8_t); // single device, single channel
    
    // .................................................... private attributes
    /// Device address as defined by pins A1, A0
    uint8_t address_;
    
    /// Array of channel objects.
    ADS7828Channel channels_[8];
    
    /// Command byte for device object (PD1 PD0 bits only).
    uint8_t commandByte_;
    
    // ............................................. static private attributes
    /// Array of pointers to registered device objects.
    static ADS7828* devices_[4];
    
    /// Factory pre-set slave address.
    static const uint8_t BASE_ADDRESS_ = 0x48;
};
#endif
/// \example examples/one_device/one_device.pde
/// \example examples/two_devices/two_devices.pde
